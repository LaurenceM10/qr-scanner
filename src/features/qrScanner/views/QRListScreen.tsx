import React, { useState } from 'react';
import {
  View,
  Text,
  FlatList,
  StyleSheet,
  SafeAreaView,
  ListRenderItem,
} from 'react-native';
import SearchInput from 'features/qrScanner/views/SearchInput';

// Styles
import { theme } from 'app/theme';

// Redux
import { useAppSelector } from 'app/store';
import { RootState } from 'app/store/store';
import { QRItem } from 'features/qrScanner/slice/types';
import NoResults from 'features/qrScanner/views/NoResults';
import NoData from 'features/qrScanner/views/NoData';

function QRListScreen() {
  const [searchTerm, setSearchTerm] = useState<string>('');
  const QRList = useAppSelector((state: RootState) => {
    const { QRData } = state.QRReducer;
    const parsedSearchTerm = searchTerm.toLowerCase();

    if (parsedSearchTerm.length) {
      return QRData.filter(
        str => str.content.toLowerCase().indexOf(parsedSearchTerm) !== -1,
      );
    }

    return QRData;
  });

  const renderItem: ListRenderItem<QRItem> = ({ item }) => (
    <View style={styles.listItem}>
      <Text style={styles.qrContent}>{item.content}</Text>
    </View>
  );

  const renderContent = () => {
    if (searchTerm.length > 0 && !QRList.length) {
      return <NoResults />;
    }

    if (!QRList.length) {
      return <NoData />;
    }

    return (
      <FlatList
        data={QRList}
        style={styles.list}
        renderItem={renderItem}
        keyExtractor={item => `row-${item.id}`}
      />
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.searchContainer}>
        <SearchInput
          autoFocus={false}
          placeholder="Search"
          onChange={setSearchTerm}
        />
      </View>
      {renderContent()}
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.background,
    paddingTop: 40,
  },
  searchContainer: {
    marginBottom: 12,
    marginHorizontal: 12,
  },
  list: {
    flex: 1,
  },
  listItem: {
    elevation: 2,
    borderRadius: 4,
    marginVertical: 4,
    paddingVertical: 12,
    marginHorizontal: 12,
    paddingHorizontal: 8,
    backgroundColor: theme.colors.background,
  },
  qrContent: {
    color: theme.colors.accent,
  },
});

export default QRListScreen;
