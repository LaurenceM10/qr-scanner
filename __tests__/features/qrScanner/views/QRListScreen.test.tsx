import React from 'react';
import { cleanup, fireEvent } from '@testing-library/react-native';

import QRListScreen from 'features/qrScanner/views/QRListScreen';
import renderWithRedux from '../../../test-utils/renderWithRedux';

const QRState = {
  QRData: [
    { id: 1, content: 'www.facebook.com' },
    { id: 2, content: 'www.youtube.com' },
    { id: 3, content: 'www.twitter.com' },
    { id: 4, content: 'www.github.com' },
  ],
};

afterEach(cleanup);

describe('<QRListScreen />', function () {
  it('should render the ReadQRScreen component when there is no permission', function () {
    const { toJSON } = renderWithRedux(<QRListScreen />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render a FlatList of QR items from the redux store', function () {
    const preloadedState = { QRReducer: QRState };
    const { queryByText } = renderWithRedux(<QRListScreen />, {
      preloadedState,
    });

    expect(queryByText(/twitter/)).toBeTruthy();
  });

  it('should filter a list of QR items based on input value', function () {
    const preloadedState = { QRReducer: QRState };
    const { getByPlaceholderText, queryByText } = renderWithRedux(
      <QRListScreen />,
      { preloadedState },
    );

    const input = getByPlaceholderText('Search');
    expect(input).toBeTruthy();

    fireEvent.changeText(input, 'twit');

    expect(queryByText(/twit/)).toBeTruthy();
    expect(queryByText(/youtube/)).toBeNull();
    expect(queryByText(/github/)).toBeNull();
  });

  it('should render <NoResults /> when no items are found based on the filter', function () {
    const preloadedState = { QRReducer: QRState };
    const { getByPlaceholderText, getByTestId, queryByText } = renderWithRedux(
      <QRListScreen />,
      { preloadedState },
    );

    const input = getByPlaceholderText('Search');
    expect(input).toBeTruthy();

    fireEvent.changeText(input, 'nicasource');

    expect(queryByText(/nicasource/)).toBeNull();
    expect(getByTestId('NoResults')).toBeTruthy();
  });

  it('should render <NoData /> when there are no QR items saved', function () {
    const preloadedState = { QRReducer: { QRData: [] } };
    const { getByTestId } = renderWithRedux(<QRListScreen />, {
      preloadedState,
    });

    expect(getByTestId('NoData')).toBeTruthy();
  });
});
