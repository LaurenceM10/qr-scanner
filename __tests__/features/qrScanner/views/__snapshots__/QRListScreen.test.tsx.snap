// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QRListScreen /> should render the ReadQRScreen component when there is no permission 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "paddingTop": 40,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 12,
        "marginHorizontal": 12,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f0f0f0",
          "borderRadius": 12,
          "flexDirection": "row",
          "paddingHorizontal": 12,
          "paddingVertical": 8,
        }
      }
    >
      <Text />
      <TextInput
        allowFontScaling={true}
        autoFocus={false}
        onChangeText={[Function]}
        placeholder="Search"
        placeholderTextColor="#888"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#333333",
            "flex": 1,
            "padding": 0,
            "top": 2,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="NoData"
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "#333333",
            "marginTop": 12,
          }
        }
      >
        There is nothing saved yet
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
